### DISCUSSION TITLE
Indenting wrapped lines

### DISCUSSION THREAD

**Discussion by KillTheMule** [fltk-rs/fltk-rs#997] [Q&A]
I have a `TextDisplay`, which wraps long lines just nicely. But I kinda wish there was a way to indent the wrapped lines (the wrapped part, that is, not the first part that fits on the first display line), but I haven't found anything that remotely seems to do that. Am I wrong? Is there a way to do that somewhat easily?

I'm pondering switching to a `HelpView` so I can arrange and color things with html just the way I want, but that will need a bit of rearranging things and would happen in the future, so I'm asking this to maybe just have this one thing done in the `TextDisplay`.

Thanks for any pointers :)

**Comment by MoAlyousef** [comment]
Hi

I'm not sure what you mean with "the wrapped part, that is, not the first part that fits on the first display line".
You can set the wrap_mode of a display widget by either bounds, pixel or column:
```rust
use fltk::{prelude::*, *};

const TXT: &str = "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.";

fn main() {
  let mut buf = text::TextBuffer::default();
  buf.set_text(TXT);
  let a = app::App::default();
  let mut win = window::Window::default().with_size(400, 300);
  let mut disp = text::TextDisplay::default_fill();
  win.end();
  win.show();

  disp.wrap_mode(text::WrapMode::AtColumn, 40);
  disp.set_buffer(buf);

  a.run().unwrap();
}
```
AtBounds disregards the second argument and just wraps at the bound of the display widget.

**Comment by KillTheMule** [comment]
Yeah, ok, I saw the wrap mode, but that doesn't seem to do what I want. Let's say I have a line "This is a line", and the display would be "This is a \nline". I'd like to specify something, let's say a number of spaces, let's say 2, such that this is displayed as "This is a \n  line". So the wrapping algorithm inserts 2 spaces after each newline it inserts.

I'm not even after something specific, maybe just a generic "indent wrapped lines" would be enough :)

**Reply by MoAlyousef** [reply]
Unfortunately there is no straightforward way to do it. You can use some of the DisplayExt methods to iterate over lines, set_insert_position, insert spaces, etc.
Another way is to wrap the Display widget inside a group which has a frametype of DownBox and is white, so it seems as if the text is indented from the border: 
```rust
use fltk::{prelude::*, *};

const TXT: &str = "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.";

fn main() {
  let mut buf = text::TextBuffer::default();
  buf.set_text(TXT);
  let a = app::App::default();
  let mut win = window::Window::default().with_size(400, 300);
  let mut grp = group::Group::default().with_size(390, 290).center_of_parent();
  let mut disp = text::TextDisplay::default().with_size(350, 250).center_of_parent();
  disp.set_frame(enums::FrameType::NoBox);
  grp.end();
  grp.set_frame(enums::FrameType::DownBox);
  grp.set_color(enums::Color::White);
  win.end();
  win.show();

  disp.wrap_mode(text::WrapMode::AtColumn, 40);
  disp.set_buffer(buf.clone());

  a.run().unwrap();
}
```
![image](https://user-images.githubusercontent.com/37966791/142670523-8baeb870-3a39-4cc7-bfd3-e3e64131c734.png)

**Reply by MoAlyousef** [reply]
The x, y coordinates can be changed of course

**Reply by KillTheMule** [reply]
Thanks for the idea, but that doesn't indent only the wrapped part, but also the non-wrapped part. It's supposed to look like

```
Lorem ipsum dolor sit amet, consectetur
    adipiscing elit, sed do eiusmod tempor 
    incididunt ut labore et dolore magna aliqua. 
```

(Supposing that's one line).

**Reply by MoAlyousef** [reply]
You can use something like the unicode_linebreak lib to simplify things:

```rust
use fltk::{prelude::*, *};
use unicode_linebreak::{linebreaks, BreakOpportunity::{Mandatory, Allowed}};

const TXT: &str = "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.";

fn main() {
  
  let breaks = linebreaks(TXT);
  let old_vec = TXT.as_bytes().to_owned();
  let mut new_vec = vec![];
  let mut indices = vec![];
  let mut iteration = 1;
  for (idx, op) in breaks {
    let cap = 40 * iteration;
    if idx > cap && (op == Allowed || op == Mandatory) {
      indices.push(idx);
      iteration += 1;
    }
  }
  for (i, _) in old_vec.iter().enumerate() {
    if indices.contains(&i) {
      new_vec.push(b'\n');
      new_vec.push(b'\t');
    }
    new_vec.push(old_vec[i]); // assuming ascii
  }

  let mut buf = text::TextBuffer::default();
  buf.set_text(&String::from_utf8_lossy(&new_vec));

  let a = app::App::default();
  let mut win = window::Window::default().with_size(400, 300);
  let mut disp = text::TextDisplay::default_fill();
  win.end();
  win.show();

  disp.set_buffer(buf);

  a.run().unwrap();
}
```
This uses a tab, but you can replace things with spaces.
![image](https://user-images.githubusercontent.com/37966791/142708541-ebbfb589-907b-475f-90ce-3772692ebef7.png)

**Reply by KillTheMule** [reply]
Hmm, can't mark that as an answer... but thanks nonetheless!
